design-toolkit/components/src/hooks/use-tree-actions/index.ts

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright 2025 Hypergiant Galactic Systems Inc. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import type { Key } from '@react-types/shared';
+import { useRef } from 'react';
+import type {
+  TreeActions,
+  TreeData,
+  TreeNode,
+  TreeNodeBase,
+  UseTreeActionsOptions,
+} from '../types';
+import {
+  assert,
+  addNodes,
+  buildLookup,
+  deleteNode,
+  getAllNodes,
+  getNode,
+  getVisibilityChange,
+  insert,
+  move,
+  moveNodes,
+  setAllNodes,
+  setNode,
+  setViewable,
+  toTree,
+  validateCache,
+} from './utils';
+
+function useIsFirstMount(): boolean {
+  const isFirst = useRef(true);
+  if (isFirst.current) {
+    isFirst.current = false;
+
+    return true;
+  }
+  return isFirst.current;
+}
+
+/**
+ * Stateless hook that transforms tree data according to actions
+ * it takes in nodes and returns a new version of the tree.
+ *
+ * Note: each operation returns the whole tree. Future iterations
+ * might want to return only the changed portion of the tree.
+ *
+ * @param nodes
+ */
+export function useTreeActions<T extends object>({
+  nodes,
+}: UseTreeActionsOptions<T>): TreeActions<T> {
+  const isFirstMount = useIsFirstMount();
+  const lastBuild = useRef<TreeNode<T>[] | null>(null);
+
+  if (isFirstMount) {
+    buildLookup<T>(nodes ?? [], new Map());
+    lastBuild.current = toTree();
+  }
+
+  validateCache<T>(nodes, lastBuild.current);
+
+  function initialize() {
+    return updateAndReturn();
+  }
+
+  function getTreeNode(key: Key): TreeNode<T> | undefined {
+    const node = getNode<T>(key);
```

Comment:
suggestion: If `getNode()` is defined as returning a `TreeNode<T>`, but internally, there is typing to force a potential `falsy` value to be returned, this can be a bug somewhere else. It really should be handled without the `as TreeNode<T>` inside `getNode()`. 

Otherwise, the check for `!node` should be needless. Have you seen edge cases where this is true?